### PR TITLE
Contributing lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py module to manage HUAWEI data center CloudEngine

### PR DESCRIPTION
**Pull request by QijunPan**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

ce_vxlan_tunnel.py
##### ANSIBLE VERSION

```
ansible 2.3.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/ansible-2.1.1.0/lib/ansible/modules/core/network/cloudengine', '/home/ansible-2.1.1.0/lib/ansible/module_utils', '/home/ansible-2.1.1.0/lib/ansible/modules', '/home/ansible-2.1.1.0', '/home/ansible-2.1.1.0/lib/ansible/utils/']
```

##### SUMMARY

```
TASK [add vni_id and bridge_domain_id] *****************************************
task path: /usr1/code/OpenNESS/code/current/KVM/intg/test/testcases/test-ce_vxlan_tunnel.yml:36
Using module file /usr/local/lib/python2.7/site-packages/ansible-2.3.0-py2.7.egg/ansible/modules/core/network/cloudengine/ce_vxlan_tunnel.py
<ce_12800> ESTABLISH LOCAL CONNECTION FOR USER: root
<ce_12800> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1487054649.63-132024260408253 `" && echo ansible-tmp-1487054649.63-132024260408253="` echo $HOME/.ansible/tmp/ansible-tmp-1487054649.63-132024260408253 `" ) && sleep 0'
<ce_12800> PUT /tmp/tmpz_zEwe TO /root/.ansible/tmp/ansible-tmp-1487054649.63-132024260408253/ce_vxlan_tunnel.py
<ce_12800> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1487054649.63-132024260408253/ /root/.ansible/tmp/ansible-tmp-1487054649.63-132024260408253/ce_vxlan_tunnel.py && sleep 0'
<ce_12800> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1487054649.63-132024260408253/ce_vxlan_tunnel.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1487054649.63-132024260408253/" > /dev/null 2>&1 && sleep 0'
changed: [ce_12800] => {
    "changed": true, 
    "end_state": {
        "vni_to_bridge_domain": [
            {
                "bdId": "20", 
                "vniId": "1"
            }
        ]
    }, 
    "existing": {}, 
    "invocation": {
        "module_args": {
            "auth_pass": null, 
            "authorize": false, 
            "bridge_domain_id": "20", 
            "host": "ce_12800", 
            "nve_mode": null, 
            "nve_name": null, 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "peer_list_ip": null, 
            "port": 12346, 
            "protocol_type": null, 
            "provider": null, 
            "source_ip": null, 
            "ssh_keyfile": null, 
            "state": "present", 
            "timeout": 10, 
            "transport": null, 
            "use_ssl": false, 
            "username": "root123", 
            "validate_certs": true, 
            "vni_id": "1"
        }, 
        "module_name": "ce_vxlan_tunnel"
    }, 
    "proposed": {
        "bridge_domain_id": "20", 
        "state": "present", 
        "vni_id": "1"
    }, 
    "updates": [
        "bridge-domain 20", 
        "vxlan vni 1"
    ]
}

```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 1]
```diff
@@ -0,0 +1,936 @@
...
+            self.results['updates'] = list()
+
+        self.module.exit_json(**self.results)
+
+
+def main():
+    """Module main"""
+
+    argument_spec = dict(
+        bridge_domain_id=dict(required=False),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:918]: `type='int'`

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 2]
```diff
@@ -0,0 +1,936 @@
...
+        vni_id=dict(required=False, type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:919]: `type='int'`

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 3]
```diff
@@ -0,0 +1,936 @@
...
+        nve_name=dict(required=False, type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:920]: `type='int'`

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 4]
```diff
@@ -0,0 +1,936 @@
...
+        self.updates_cmd.append(
+            "undo vni %s head-end peer-list protocol bgp " % vni_id)
+        self.changed = True
+
+    def check_params(self):
+        """Check all input params"""
+
+        # bridge_domain_id check
+        if self.bridge_domain_id:
+            if not self.bridge_domain_id.isdigit():
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:767]: specifying `type='int'` removes the need for this, so you only need to check the range.

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 5]
```diff
@@ -0,0 +1,936 @@
...
+                self.module.fail_json(
+                    msg='Error: The parameter of bridge domain id is invalid.')
+            if int(self.bridge_domain_id) > 16777215 or int(self.bridge_domain_id) < 1:
+                self.module.fail_json(
+                    msg='Error: The bridge domain id must be an integer between 1 and 16777215.')
+        # vni_id check
+        if self.vni_id:
+            if not self.vni_id.isdigit():
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:775]: specifying `type='int'` removes the need for this, so you only need to check the range.

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 6]
```diff
@@ -0,0 +1,936 @@
...
+
+        self.changed = True
+
+    def config_delete_mode(self, nve_name, mode):
+        """nve mode"""
+
+        if mode == "mode-l3":
+            if not self.is_nve_mode_exist(nve_name, mode):
+                return
+            mode = "mode-l2"
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py]: I'm not familiar with CE, though is this correct?
```
if mode == "mode-l3":
   ...
   mode = "mode-l2"
   cfg_xml = CE_NC_MERGE_NVE_MODE % (nve_name, mode)
   ...
   self.updates_cmd.append("undo mode l3")
```

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 7]
```diff
@@ -0,0 +1,936 @@
...
+        self.updates_cmd.append("interface %s" % nve_name)
+        self.updates_cmd.append("undo source %s" % source_ip)
+        self.changed = True
+
+    def config_delete_vni_peer_ip(self, nve_name, vni_id, peer_ip_list):
+        """remove vni peer ip"""
+
+        for peer_ip in peer_ip_list:
+            if not self.is_vni_peer_list_exist(nve_name, vni_id, peer_ip):
+                self.module.fail_json(msg='Error: The %s is not exist' % peer_ip)
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py]: Perhaps say `does not exist`?

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 8]
```diff
@@ -0,0 +1,936 @@
...
+                    msg='Error: The parameter of vni id is invalid.')
+            if int(self.vni_id) > 16000000 or int(self.vni_id) < 1:
+                self.module.fail_json(
+                    msg='Error: The vni id must be an integer between 1 and 16000000.')
+
+        # nve_name check
+        if self.nve_name:
+            if not self.check_nve_name():
+                self.module.fail_json(
+                    msg='Error: Error: Nve interface is invalid.')
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py]: Perhaps include the name of the NVE interface in the error message?
Also `Nve` or `NVE`?

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py [hunk 9]
```diff
@@ -0,0 +1,936 @@
...
+
+        # peer_list_ip check
+        if self.peer_list_ip:
+            for peer_ip in self.peer_list_ip:
+                if not is_valid_address(peer_ip):
+                    self.module.fail_json(
+                        msg='Error: The %s exist invalid ip address' % self.peer_list_ip)
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py]: Perhaps: `Error: The %s exist, and is an invalid IP address`?

**Review by gundalow — CHANGES_REQUESTED** [review]
I've added a note of the Core Team Agenda to see if we can add built in validation of int ranges and IP addresses as part of `argument_spec`

https://github.com/ansible/community/issues/159#issuecomment-298523500

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py (hunk 6 above)
Comment (QijunPan) [inline lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py]: Keep the string type and avoid the repeated conversion, thanks!

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py (hunk 4 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:767]: Keep the string type and avoid the repeated conversion, thanks!

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py (hunk 5 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:775]: Keep the string type and avoid the repeated conversion, thanks!

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py (hunk 7 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py]: yes, thank you. I'll fix it.

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py (hunk 1 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:918]: Keep the string type and avoid the repeated conversion, thanks!

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py (hunk 2 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:919]: Keep the string type and avoid the repeated conversion, thanks!

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py (hunk 3 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py:920]: Keep the string type and avoid the repeated conversion, thanks!

File: lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py (hunk 6 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_vxlan_tunnel.py]: This is correct, but easy to cause misunderstanding, I'll fix it. thank you.

**Comment by QijunPan** [comment]
@gundalow   These problems have been fixed， can get this PR merged?

**Review by gundalow — APPROVED** [review]

**Pull request merged**